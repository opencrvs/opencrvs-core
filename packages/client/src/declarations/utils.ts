/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { EVENT_STATUS } from '@client/workqueue'
import { SUBMISSION_STATUS } from '.'

export const isPendingCorrection = (status: SUBMISSION_STATUS | undefined) =>
  status === EVENT_STATUS.CORRECTION_REQUESTED

export const isReviewableDeclaration = (
  status: SUBMISSION_STATUS | undefined
) => status && [EVENT_STATUS.DECLARED, EVENT_STATUS.VALIDATED].includes(status)

export const isUpdatableDeclaration = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.DRAFT,
    EVENT_STATUS.IN_PROGRESS,
    EVENT_STATUS.REJECTED
  ].includes(status)

export const isPrintable = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [SUBMISSION_STATUS.REGISTERED, SUBMISSION_STATUS.ISSUED].includes(status)

export const isIssuable = (status: SUBMISSION_STATUS | undefined) =>
  status === SUBMISSION_STATUS.CERTIFIED

export const isRecordOrDeclaration = (status: SUBMISSION_STATUS | undefined) =>
  status
    ? [
        SUBMISSION_STATUS.REGISTERED,
        SUBMISSION_STATUS.CORRECTION_REQUESTED,
        SUBMISSION_STATUS.CERTIFIED
      ].includes(status)
      ? 'record'
      : 'declaration'
    : ''

export const canBeCorrected = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.REGISTERED,
    SUBMISSION_STATUS.CERTIFIED,
    SUBMISSION_STATUS.ISSUED
  ].includes(status)

export const isArchivable = (status: SUBMISSION_STATUS | undefined) =>
  status &&
  [
    SUBMISSION_STATUS.IN_PROGRESS,
    SUBMISSION_STATUS.DECLARED,
    SUBMISSION_STATUS.VALIDATED,
    SUBMISSION_STATUS.REJECTED
  ].includes(status)

export const canBeReinstated = (status: SUBMISSION_STATUS | undefined) =>
  status === SUBMISSION_STATUS.ARCHIVED

export const isViewable = (status: SUBMISSION_STATUS | undefined) =>
  status !== SUBMISSION_STATUS.DRAFT
