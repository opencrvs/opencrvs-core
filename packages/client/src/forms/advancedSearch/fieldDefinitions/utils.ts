/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
// eslint-disable-next-line no-restricted-imports
import { isValid, parse } from 'date-fns'

export enum TIME_PERIOD {
  LAST_7_DAYS = 'LAST_7_DAYS',
  LAST_30_DAYS = 'LAST_30_DAYS',
  LAST_90_DAYS = 'LAST_90_DAYS',
  LAST_YEAR = 'LAST_YEAR'
}

const validateDate = (dateString: string) => {
  const formats = ['yyyy-M-d', 'yyyy-MM-dd']
  return formats.some((format) =>
    isValid(parse(dateString, format, new Date()))
  )
}

export function isInvalidDate(date?: string) {
  if (!date) return true
  const regEx = /^\d{4}-\d{1,2}-\d{1,2}$/
  if (!date.match(regEx)) {
    return true
  }
  return !validateDate(date)
}
