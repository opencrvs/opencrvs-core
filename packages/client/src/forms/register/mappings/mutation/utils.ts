/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import type { GQLRelatedPersonInput } from '@client/utils/gateway-deprecated-do-not-use'
import { ICertificate, IFileValue } from '@client/forms'
import { omit } from 'lodash'

export function stripTypename(obj: any): any {
  if (Array.isArray(obj)) {
    return obj.map(stripTypename)
  } else if (obj !== null && typeof obj === 'object') {
    const newObj: any = {}
    for (const key in obj) {
      if (key !== '__typename' && Object.hasOwn(obj, key)) {
        newObj[key] = stripTypename(obj[key])
      }
    }
    return newObj
  }
  return obj
}
export function transformCertificateData(certificates: ICertificate[]) {
  const certificateData = stripTypename(certificates[0])

  // Prepare the base certificate data
  const updatedCertificates: ICertificate[] = [
    {
      ...omit(certificateData, 'collector')
    }
  ]

  // for collector mapping
  if (certificateData && certificateData.collector) {
    let collector: GQLRelatedPersonInput = {}
    if (certificateData.collector.type) {
      collector.relationship = certificateData.collector.type as string
    }
    if (certificateData.collector.relationship) {
      collector.otherRelationship = certificateData.collector
        .relationship as string
      collector = {
        ...collector,
        name: [
          {
            use: 'en',
            firstNames: certificateData.collector.firstName as string,
            familyName: certificateData.collector.lastName as string
          }
        ],
        identifier: [
          {
            id: certificateData.collector.iD as string,
            type: certificateData.collector.iDType as string
          }
        ]
      }
    }
    if (certificateData.collector.affidavitFile) {
      collector.affidavit = [
        {
          contentType: (certificateData.collector.affidavitFile as IFileValue)
            .type,
          data: (certificateData.collector.affidavitFile as IFileValue).data
        }
      ]
    }
    updatedCertificates[0].collector = collector as any
  }

  // Return the processed certificates array
  return updatedCertificates
}
