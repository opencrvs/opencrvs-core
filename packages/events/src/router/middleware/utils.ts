/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import type { AnyTRPCMiddlewareFunction } from '@trpc/server'
import {
  ActionInputWithType,
  TokenWithBearer,
  TokenUserType
} from '@opencrvs/commons'

export interface UserDetails {
  type: TokenUserType.USER | TokenUserType.SYSTEM
  id: string
  primaryOfficeId: string | undefined
  role: string
}

export type Context = { token: TokenWithBearer; user: UserDetails }
/**
 * TRPC Middleware options with correct context.
 * Actual middleware type definition is only for internal use within TRPC.
 */
export type MiddlewareOptions = Omit<
  Parameters<AnyTRPCMiddlewareFunction>[0],
  'ctx'
> & { ctx: Context }

export type ActionMiddlewareOptions = Omit<MiddlewareOptions, 'input'> & {
  input: ActionInputWithType
}
