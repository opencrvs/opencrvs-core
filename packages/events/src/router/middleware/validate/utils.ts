/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { TRPCError } from '@trpc/server'
import _ from 'lodash'
import {
  ActionUpdate,
  FieldConfig,
  getFieldValidationErrors,
  Inferred,
  errorMessages
} from '@opencrvs/commons/events'

type ValidationError = {
  message: string
  id: string
  value: unknown
}

export function getFormFieldErrors(formFields: Inferred[], data: ActionUpdate) {
  return formFields.reduce(
    (errorResults: ValidationError[], field: FieldConfig) => {
      const fieldErrors = getFieldValidationErrors({
        field,
        values: data
      }).errors

      if (fieldErrors.length === 0) {
        return errorResults
      }

      // For backend, use the default message without translations.
      const errormessageWithId = fieldErrors.map((error) => ({
        message: error.message.defaultMessage,
        id: field.id,
        value: data[field.id as keyof typeof data]
      }))

      return [...errorResults, ...errormessageWithId]
    },
    []
  )
}

export function getVerificationPageErrors(
  verificationPageIds: string[],
  data: ActionUpdate
) {
  return verificationPageIds
    .map((pageId) => {
      const value = data[pageId]
      return typeof value !== 'boolean'
        ? {
            message: 'Verification page result is required',
            id: pageId,
            value
          }
        : null
    })
    .filter((error) => error !== null)
}

export function throwWhenNotEmpty(errors: unknown[]) {
  if (errors.length > 0) {
    throw new TRPCError({
      code: 'BAD_REQUEST',
      message: JSON.stringify(errors)
    })
  }
}

function isObject(value: unknown): value is Record<string, unknown> {
  return _.isPlainObject(value)
}

/**
 * Flattens an object into a list of [key, value] pairs using dot/bracket notation.
 */
function flattenEntries<T>(obj: T, path: string = ''): [string, unknown][] {
  if (Array.isArray(obj)) {
    return obj.flatMap((value, index) =>
      flattenEntries(value, `${path}[${index}]`)
    )
  }

  if (isObject(obj)) {
    return Object.entries(obj).flatMap(([key, value]) =>
      flattenEntries(value, path ? `${path}.${key}` : key)
    )
  }

  return obj === undefined ? [] : [[path, obj]]
}

/**
 * Compares update vs cleaned payloads and returns an array of offending keys.
 */
export function getInvalidUpdateKeys<T>({
  update,
  cleaned
}: {
  update: T
  cleaned: T
}): ValidationError[] {
  const updateEntries = flattenEntries(update)
  const cleanedKeys = flattenEntries(cleaned).map(([key]) => key)

  return updateEntries
    .filter(([key]) => !cleanedKeys.includes(key))
    .map(([key, value]) => ({
      message: errorMessages.hiddenField.defaultMessage,
      id: key,
      value
    }))
}
