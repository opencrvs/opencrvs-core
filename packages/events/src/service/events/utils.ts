/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { Action, ActionType } from '@opencrvs/commons'

export function findLastAssignmentAction(actions: Action[]) {
  return actions
    .filter(
      ({ type }) => type === ActionType.ASSIGN || type === ActionType.UNASSIGN
    )
    .reduce<
      Action | undefined
    >((latestAction, action) => (!latestAction || action.createdAt > latestAction.createdAt ? action : latestAction), undefined)
}
