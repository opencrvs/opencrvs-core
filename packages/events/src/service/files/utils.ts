/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { differenceWith, first, isEqual, orderBy } from 'lodash'
import {
  Draft,
  EventConfig,
  EventDocument,
  FieldConfig,
  FileFieldType,
  getActionAnnotationFields,
  AnyFileField,
  getDeclarationFields,
  getMixedPath,
  FileFieldWithOptionValue,
  FileFieldValue,
  ActionUpdate,
  getCurrentEventStateWithDrafts,
  EventState,
  getCurrentEventState,
  ActionDocument
} from '@opencrvs/commons'

function isFieldWithFile(field: FieldConfig): field is AnyFileField {
  return FileFieldType.safeParse(field.type).success
}

/**
 *
 * @param field File field
 * @param update single action or event state.
 * @returns normalised file reference based on the field id and event state.
 */
function getFileReferences(
  field: AnyFileField,
  update: ActionUpdate | EventState
): { id: string; value: string }[] {
  const value = getMixedPath<AnyFileField['defaultValue']>(update, field.id)

  if (!value) {
    return []
  }

  const maybeFileOptions = FileFieldWithOptionValue.safeParse(value)
  if (maybeFileOptions.success) {
    return maybeFileOptions.data.map((opt) => ({
      id: field.id,
      value: opt.path
    }))
  }

  const maybeFile = FileFieldValue.safeParse(value)

  if (maybeFile.success) {
    return [{ id: field.id, value: maybeFile.data.path }]
  }

  if (typeof value === 'string') {
    return [{ id: field.id, value }]
  }

  throw new Error(
    `Unknown filetype, ${field.type} with value: ${JSON.stringify(value)}`
  )
}

/**
 *  Since API allows sending PATCH requests with only the changed fields,
 *  we need to check the current event state with the current draft.
 * @returns file references that are not present in the current event state but are when previous draft is applied to the state.
 */
function getUnreferencedDeclarationDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const currentEventState = currentDraft
    ? getCurrentEventStateWithDrafts({
        event,
        draft: currentDraft,
        configuration
      })
    : getCurrentEventState(event, configuration)

  const previousEventState = getCurrentEventStateWithDrafts({
    event,

    draft: {
      // Given previous draft, if event has been updated afterwards (running deletion after submission of actual action.) the data would never get applied.
      ...previousDraft,
      createdAt: new Date().toISOString(),
      action: {
        ...previousDraft.action,
        createdAt: new Date().toISOString()
      }
    },
    configuration
  })

  const declarationFileFields =
    getDeclarationFields(configuration).filter(isFieldWithFile)

  const currentDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, currentEventState.declaration)
  )

  const previousDeclarationFiles = declarationFileFields.flatMap((field) =>
    getFileReferences(field, previousEventState.declaration)
  )

  return differenceWith(
    previousDeclarationFiles,
    currentDeclarationFiles,
    isEqual
  )
}

function getUnreferencedAnnotationDraftFiles({
  currentAction,
  previousDraft,
  configuration
}: {
  currentAction: Draft['action'] | ActionDocument
  previousDraft: Draft
  configuration: EventConfig
}) {
  if (currentAction.type !== previousDraft.action.type) {
    // eslint-disable-next-line no-console
    console.warn(
      'Previous draft was for different action. Not comparing file references between them.'
    )
    return []
  }

  const currentDraftActionConfig = configuration.actions.find(
    (action) => action.type === previousDraft.action.type
  )

  if (!currentDraftActionConfig) {
    // eslint-disable-next-line no-console
    console.warn(
      `No configuration found for action ${previousDraft.action.type}.`
    )
  }

  const annotationFileFields = currentDraftActionConfig
    ? getActionAnnotationFields(currentDraftActionConfig).filter(
        isFieldWithFile
      )
    : []

  // @TODO: There might be a case where annotation shares id with declaration field? Should it limited by the configuration?
  const currentAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, currentAction.annotation ?? {})
  )

  const previousAnnotationFiles = annotationFileFields.flatMap((field) =>
    getFileReferences(field, previousDraft.action.annotation ?? {})
  )

  return differenceWith(
    previousAnnotationFiles,
    currentAnnotationFiles,
    isEqual
  )
}

/**
 *
 * @returns all the filenames that are referenced in the current draft, but not in the previous draft.
 */
export function getUnreferencedDraftFiles({
  event,
  currentDraft,
  previousDraft,
  configuration
}: {
  event: EventDocument
  /**
   * Current draft is optional, because we might be comparing the previous draft with the current event state. (e.g. actual action has been submitted)
   */
  currentDraft?: Draft
  previousDraft: Draft
  configuration: EventConfig
}) {
  const unreferencedDeclarationDraftFiles =
    getUnreferencedDeclarationDraftFiles({
      event,
      currentDraft,
      previousDraft,
      configuration
    })

  const currentAction =
    currentDraft?.action ??
    first(
      orderBy(
        event.actions.filter(
          (action) => action.type === previousDraft.action.type
        ),
        ['createdAt'],
        'desc'
      )
    )

  if (!currentAction) {
    // This should not happen. It means that the previous draft action type was invalid.
    // eslint-disable-next-line no-console
    console.error(
      `Could not find corresponding action ${previousDraft.action.type} for event: ${event.id}`
    )

    return unreferencedDeclarationDraftFiles
  }

  const unreferencedAnnotationDraftFiles = getUnreferencedAnnotationDraftFiles({
    // @TODO: for some reason declaration is omitted from some type?
    currentAction: currentAction as Draft['action'] | ActionDocument,
    previousDraft,
    configuration
  })

  return [
    ...unreferencedDeclarationDraftFiles,
    ...unreferencedAnnotationDraftFiles
  ]
}
