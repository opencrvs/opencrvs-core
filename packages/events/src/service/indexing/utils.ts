/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */
import { EventSearchIndex } from '@opencrvs/commons/events'
import {
  QueryDslQueryContainer,
  SearchRequest
} from '@elastic/elasticsearch/lib/api/types'

const FIELD_SEPARATOR = '____'
export const DEFAULT_SIZE = 10

/**
 * Generates an Elasticsearch query to search within `document.data`
 * using the provided search payload.
 */
export function generateQuery(event: Omit<EventSearchIndex, 'type'>) {
  const must: QueryDslQueryContainer[] = Object.entries(event).map(
    ([key, value]) => ({
      match: {
        [`data.${key.replaceAll('.', FIELD_SEPARATOR)}`]: value
      }
    })
  )

  return {
    bool: {
      must
    }
  } as SearchRequest['query']
}
