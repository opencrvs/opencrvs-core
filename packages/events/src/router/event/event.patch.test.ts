/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * OpenCRVS is also distributed under the terms of the Civil Registration
 * & Healthcare Disclaimer located at http://opencrvs.org/license.
 *
 * Copyright (C) The OpenCRVS Authors located at https://github.com/opencrvs/opencrvs-core/blob/master/AUTHORS.
 */

import { createTestClient, setupTestCase } from '@events/tests/utils'

test('Prevents updating event with unknown type', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const event = await client.event.create(generator.event.create())

  await expect(
    client.event.patch(
      generator.event.patch(event.id, {
        type: 'EVENT_TYPE_THAT_DOES_NOT_EXIST'
      })
    )
  ).rejects.toThrowErrorMatchingSnapshot()
})

test('stored events can be modified', async () => {
  const { user, generator } = await setupTestCase()
  const client = createTestClient(user)

  const originalEvent = await client.event.create(generator.event.create())

  const event = await client.event.patch(
    generator.event.patch(originalEvent.id, {
      type: 'TENNIS_CLUB_MEMBERSHIP_PREMIUM'
    })
  )

  expect(event.updatedAt).not.toBe(originalEvent.updatedAt)
  expect(event.type).toBe('TENNIS_CLUB_MEMBERSHIP_PREMIUM')
})
