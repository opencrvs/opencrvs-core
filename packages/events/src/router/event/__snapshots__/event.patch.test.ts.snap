// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Prevents updating event with unknown type 1`] = `[TRPCError: Invalid event type EVENT_TYPE_THAT_DOES_NOT_EXIST. Valid event types are: TENNIS_CLUB_MEMBERSHIP, TENNIS_CLUB_MEMBERSHIP_PREMIUM]`;
